include/idf/EthernetDevice.hh

```diff
@@ -0,0 +1,164 @@
+/*
+PURPOSE:
+LIBRARY DEPENDENCIES: (
+(idf/EthernetDevice.cpp)
+)
+*/
+
+/**
+ * @trick_parse{everything}
+ * @trick_link_dependency{idf/EthernetDevice.cpp}
+ */
+
+#ifndef ETHERNET_DEVICE_HH
+#define ETHERNET_DEVICE_HH
+
+#include "idf/InputDevice.hh"
+
+#include <string>
+#include <netinet/in.h>
+
+namespace idf {
+
+/**
+ * @brief abstract base class for all ethernet input devices. Subclasses should usually
+ * represet a specific physical device and implement logic to fulfill the contract of
+ * open(). They should also usually derive from the appropriate class from the %Input
+ * Abstration Layer, as well as be concrete.
+ * 
+ * @author  Philip Kunz
+ */
+class EthernetDevice : public InputDevice {
+
+    public:
+
+    /**
+     * @brief constructs a new instance whose open() will open the UDP port at
+     * @a ipAddress: @a port
+     * 
+     * @param ipAddress the ipaddress of this device
+     * @param port the port number of this device
+     */
+    EthernetDevice(const std::string& name, unsigned packetLength);
+
+    virtual ~EthernetDevice() {
+        close();
+    };
+    
+    virtual void open();
+    virtual void close();
+
+    /**
+     * sets the host to which this instance will attempt to connect the next
+     * time open() is called
+     *
+     * @param host @copydoc serverName
+     */
+    void setHost(const std::string& host) {
+        serverName = host;
+    }
+
+    /**
+     * sets the server port to which this instance will attempt to connect the
+     * next time open() is called
+     *
+     * @param port @copydoc serverPort
+     */
+    void setPort(unsigned short port) {
+        serverPort = port;
+    }
+
+    /**
+     * @brief Sets the communications protocol to TCP.
+     * default is TCP
+     */
+    void setTCP() {
+        tcp = true;
+        srcAddr = NULL;
+        srcAddrLen = 0;
+        sockType = SOCK_STREAM;
+    }
+
+    /**
+     * @brief Sets the communications protocol to UDP.
+     * default is TCP.
+     * @param greeting message to send to server when connecting
+     */
+    void setUDP(const char* greeting) {
+        tcp = false;
+        sockType = SOCK_DGRAM;
+        udpGreeting = greeting;
+    }
+
+    protected:
+
+    virtual std::vector<std::vector<unsigned char> > read();
+
+    /**
+     * reads @a length bytes from this device and stores them in @a buffer
+     *
+     * @param buffer the location to store the data
+     * @param length the number of bytes to read
+     *
+     * @return the number of bytes read (always non-negative)
+     *
+     * @throws IOException if an error occurs while reading or if the device is not open
+     */
+    virtual int read(unsigned char *buffer, size_t length);
```

Comment:
the comments for the return values for these I/O calls says the return is always non-negative and these aren't overriding a base-class method I believe. Can we just return the standard `size_t` int type? for read, peek, and write?